src/Qir/Runtime/lib/QIR/QubitManagerRestrictedReuse.hpp

```diff
@@ -0,0 +1,210 @@
+        public: ~CQubitManagerRestrictedReuse();
+        private: void EnsureCapacity(int requestedCapacity);
+
+        // Disable, disabled qubit count and checks
+
+        public: int DisabledQubitsCount = 0;
+        public: bool IsDisabled(int id);
+        public: bool IsDisabled(Qubit qubit);
+        // Disables a given qubit.
+        // Once a qubit is disabled it can never be "enabled" or reallocated.
+        public: void Disable(Qubit qubit);
+        // Disables a set of given qubits.
+        // Once a qubit is disabled it can never be "enabled" or reallocated.
+        public: void Disable(Qubit* qubitsToDisable, int qubitCount);
+
+        // Allocate, allocated qubit count and checks
+
+        public: int AllocatedQubitsCount = 0;
+        public: bool IsExplicitlyAllocated(long id);
+        public: void ChangeStatusToAllocated(long id);
+        // Allocates a qubit. Extend capacity if necessary and possible.
+        // Fails if the qubit cannot be allocated. 
+        public: Qubit Allocate();
+        // Allocates numToAllocate new qubits. Extend capacity if necessary and possible.
+        // Fails without allocating any qubits if the qubits cannot be allocated. 
+        // Returned array of qubits must be passed to Release to release qubits and free memory use for array.
+        public: Qubit* Allocate(long numToAllocate);
+
+        // Release, free qubit count and checks
+
+        public: long FreeQubitsCount = 0;
+        public: bool IsFree(long id);
+        public: bool IsFree(Qubit qubit);
+        public: void Release(long id);
+        // Releases a given qubit.
+        public: void Release(Qubit qubit);
+        // Releases an array of given qubits and deallocates array memory.
+        public: void Release(Qubit* qubitsToRelease, int qubitCount);
+
+        // Borrow (We treat borrowing as allocation currently)
+
+        public: Qubit Borrow();
+        public: Qubit* Borrow(long qubitCountToBorrow);
+
+        // Return (We treat returning as release currently)
+
+        public: void Return(Qubit qubit);
+        public: void Return(Qubit* qubitsToReturn, int qubitCountToReturn);
+
+        // Qubit creation, validity check
+
+        public: bool IsValid(Qubit qubit);
+
+        // May be overriden to create a custom Qubit object.
+        // When not overriden, it just stores qubit Id in place of a pointer to a qubit.
+        // id: unique qubit id
+        // Returns a newly instantiated qubit.
+        public: virtual Qubit CreateQubitObject(int id);
+
+        // May be overriden to get a qubit id from a custom qubit object.
+        // Must be overriden if CreateQubitObject is overriden.
+        // When not overriden, it just reinterprets pointer to qubit as a qubit id.
+        // qubit: pointer to QUBIT
+        // Returns id of a qubit pointed to by qubit.
+        public: virtual int QubitToId(Qubit qubit);
+    };
```

Comment:
Consider grouping the members by access (`public` first, then `protected`, then `private`):
```c++
class C ..
{
public:
    ..
protected:
    ..
private:
    ..
};
```
I understand that sometimes it creates code that cannot be compiled in a single parse (e.g. the member variables are mentioned in the `public` constructor first, and then they are defined in the `private` section). But today's compilers are multi-parse. And that's the approach many people use (list the `public` first).

Reply:
This is not about a single pass compilation. This is about C# heritage. In C# each member carries access rights. So the members were grouped according to similar functionality. I.e. everything that concerns qubit release were in one region, etc. Yes it's more common in C++ to group members by access rights and I changed the code. This way it became a bit more convenient for _users_ of this class rather than _implementors_.